### DISCUSSION TITLE
MultiSig Plugin in Modular Account

### DISCUSSION THREAD

**Discussion by AbhishekCDN** [alchemyplatform/aa-sdk#865] [Issues] [answered]
Hey, I am creating a Modular Account through Alchemy AA SDK. After the creation of Modular account I want to extend that account with installing multiSig pulugin in it. But I'm getting a Error :- 
![ModularMultiSig](https://github.com/user-attachments/assets/689e9a35-b515-4930-a442-3b50c9d1bee1)
![Error](https://github.com/user-attachments/assets/38229f6a-3971-472c-9ed7-de8276c3ca72)

**Comment by HatimPatwa** [comment]
Same Here ,also can we get a reply over here in this thread,

**Comment by avarobinson** [comment]
Hey @HatimPatwa @AbhishekCDN!

Just want to clarify - are you trying to use the Multi-owner plugin or the Muli-sig plugin?

Multiowner will allow for full equal control of the account for multiple owners, where as multi-sig will add a `k of n` threshold signature scheme to your account. 

In the example you shared @AbhishekCDN, it looks like you are trying to add full owner control (with the 1n threshold). The easier way to do this would be to use the Multiowner plugin which is installed in Modular Account by default. This [guide](https://accountkit.alchemy.com/using-smart-accounts/transfer-ownership/multi-owner-light-account#2-add-or-remove-owners-for-a-multiownerlightaccount) walks through managing ownership. 

If alternatively, you do want to use multi-sig, I would recommend you use the multi-sig account client documented [here](https://accountkit.alchemy.com/smart-accounts/modular-account/multisig-plugin/). 

Let me know which functionality you are looking for and I'll provide the easiest solution for you.

**Reply by AbhishekCDN** [reply]
Hey @avarobinson 

I want to use the [Multi-sig plugin](https://accountkit.alchemy.com/smart-accounts/modular-account/#multisig-plugin).  Which is extend on the Modular Account

In our application we create a modular account then if a user wants this account as a multi-sig account at that time we want to add a multi-sig plugin to our modular account.

**Comment by avarobinson** [comment, marked as answer]
@AbhishekCDN @HatimPatwa Thanks for your patience! For context, this use case is not currently easily supported with the SDK. We are now working on multiple improvements to improve this workflow in the next version of the SDK! 

**Recommended flow today**
Rather than switch ownership post creation, use `createModularAccountAlchemyClient` OR the `createMultisigAccountAlchemyClient` on account creation depending on if you want multi-owner or multi-sig ownership respectively.

**Problem**
In order to switch from multi-owner (installed in Modular Account by default) to multi-sig, you will need to batch call an uninstall of the multi-owner plugin and an install of the multi-sig plugin. **It is required to batch these steps!** If you uninstall the multi-owner plugin in a single action, the account will no longer have a validation function, and will be bricked (unusable). 

**Solution for extending Modular Account AFTER deployment** 
You will need to manually encode the uninstall of the multi-owner and install of multi-sig and batch those together in one UO. 

```
// 1. Create a normal modular alchemy account
const smartAccountClient = await createModularAccountAlchemyClient({
    apiKey: API_KEY,
    chain: chain,
    signer: signer,
    gasManagerConfig: { policyId: GAS_MANAGER_POLICY_ID },
});
console.log("sca address", smartAccountClient.getAddress());

// 2. Batch execute the uninstall of multi-owner plugin and install of multi-sig plugin
// 2a. encode uninstall multiowner plugin
const multiOwnerPluginAddress = MultiOwnerPlugin.meta.addresses[
    chain.id
  ] as Address;

const encodedUninstallMultiownerPluginUserOp = encodeFunctionData({
    abi: IPluginManagerAbi,
    functionName: "uninstallPlugin",
    args: [multiOwnerPluginAddress, "0x", "0x"],
});

// 2b. encode install multi-sig plugin
const multiSigPluginAddress = MultisigPlugin.meta.addresses[
    chain.id
  ] as Address;
 
const pluginManifest = await smartAccountClient.readContract({
    abi: IPluginAbi,
    address: multiSigPluginAddress,
    functionName: "pluginManifest",
});

const manifestHash: Hash = keccak256(
    encodeFunctionResult({
      abi: IPluginAbi,
      functionName: "pluginManifest",
      result: pluginManifest,
    })
);

// TODO: define your own multi-sig parameters here (i.e. list of signer addresses and threshold requirements)
const signers = [signer, signer2];
const owners = await Promise.all(signers.map((s) => s.getAddress()));
const threshold = 1n;
const multiSigPluginInitData = encodeAbiParameters(
    [{ type: "address[]" }, { type: "uint256" }],
    [owners, threshold]
);

const encodedInstallMultiSigPluginUserOp = encodeFunctionData({
    abi: IPluginManagerAbi,
    functionName: "installPlugin",
    args: [multiSigPluginAddress, manifestHash, multiSigPluginInitData, []],
});

// 2c. Batch user op together
const { hash } = await smartAccountClient.sendUserOperation({
    uo: [
      {
        target: await smartAccountClient.getAddress(),
        data: encodedUninstallMultiownerPluginUserOp,
      },
      {
        target: await smartAccountClient.getAddress(),
        data: encodedInstallMultiSigPluginUserOp,
      },
    ],
  });
console.log("uo hash", hash);

// Wait for UO to land
const response = await smartAccountClient.waitForUserOperationTransaction({
    hash,
});
console.log("uninstalled and installed successfully!", response);

// 3. confirm installation
const accountLoupeActionsExtendedClient =
    smartAccountClient.extend(accountLoupeActions);

// returns addresses of all installed plugins
const installedPlugins =
    await accountLoupeActionsExtendedClient.getInstalledPlugins({});
console.log("installed plugins", installedPlugins); // Can verify on etherscan to ensure the installed plugin is multi-sig
```

**Reply by avarobinson** [reply]
Note that after successfully changing the ownership model to multi-sig, you will need to use the multi-sig specific client to interact with the account and send user operations. You can then follow [this](https://accountkit.alchemy.com/smart-accounts/modular-account/multisig-plugin/getting-started) guide for sending and signing UOs with multi-sig.

Example:
```
const smartAccountClientMultiSig = await createMultisigAccountAlchemyClient({
    chain,
    signer,
    owners,
    threshold,
    apiKey: API_KEY,
    accountAddress: "SCA ADDRESS", // get this from the previous modular account client (smartAccountClient.getAddress())
    gasManagerConfig: { policyId: GAS_MANAGER_POLICY_ID },
});
```

**Reply by UMainLove** [reply]
I was wondering how to use both plugins instead of switching from the Multi-Owner to the Multi-sig. Is it possible? 

This question could be extended in a wider design pattern where a brand new plugin could be added to a msca list of installed plugins and used whenever users want leveraging even the functionalities of other already installed plugins but still acting as a singleton...

**Comment by AbhishekCDN** [comment]
@avarobinson  Thanks for the solution. Will try to implement this and LYK.
